src/rrules/avoiding_non_differentiable_code.jl

```diff
@@ -14,25 +14,151 @@ end
 @zero_adjoint MinimalCtx Tuple{Type{Float16},Any,RoundingMode}
 @zero_adjoint MinimalCtx Tuple{typeof(==),Type,Type}
 
+# logging, String related primitive rules
+using Base: getindex, getproperty
+using Base.Threads: Atomic
+using Mooncake: zero_fcodual, MinimalCtx, @is_primitive, NoPullback, CoDual
+using Base.CoreLogging: LogLevel, handle_message, invokelatest
+import Base.CoreLogging as CoreLogging
+
+# Rule for accessing an Atomic{T} wrapped Integer with Base.getindex as deriving a rule results
+# in encountering a Atomic->Int address bitcast followed by a llvm atomic load call 
+@zero_adjoint MinimalCtx Tuple{typeof(getindex),Atomic{I}} where {I<:Integer}
+
+# Some Base String related rrules :
+@zero_adjoint MinimalCtx Tuple{typeof(println),Vararg}
+@zero_adjoint MinimalCtx Tuple{typeof(print),Vararg}
+@zero_adjoint MinimalCtx Tuple{typeof(show),Vararg}
+@zero_adjoint MinimalCtx Tuple{typeof(normpath),String}
+
+# seperate kwargs, non-kwargs Base.sprint rules are required. Julia compilation only gives a common lowered IR for any Base.sprint calls.
+# refer issue #558 and PR https://github.com/chalk-lab/Mooncake.jl/pull/659 for another sneaky appearance of this problem + fix.
+@zero_adjoint MinimalCtx Tuple{typeof(sprint),Vararg}
+@is_primitive MinimalCtx Tuple{typeof(Core.kwcall),<:NamedTuple,typeof(sprint),Vararg}
+function rrule!!(
+    ::CoDual{typeof(Core.kwcall)},
+    kwargs::CoDual{<:NamedTuple},
+    ::CoDual{typeof(sprint)},
+    args::Vararg{CoDual},
+)
+    primal_args = map(x -> x.x, args)
+    result = Core.kwcall(kwargs.x, sprint, primal_args...)
+    return zero_fcodual(result),
+    NoPullback(zero_fcodual(Core.kwcall), kwargs, zero_fcodual(sprint), args...)
+end
+
+# Base.CoreLogging @logmsg related primitives.
+@zero_adjoint MinimalCtx Tuple{
+    typeof(Base._replace_init),String,Tuple{Pair{String,String}},Int64
+}
+@zero_adjoint MinimalCtx Tuple{
+    typeof(CoreLogging.current_logger_for_env),LogLevel,Symbol,Module
+}
+@zero_adjoint MinimalCtx Tuple{
+    typeof(Core._call_latest),
+    typeof(Base.CoreLogging.shouldlog),
+    Any,
+    LogLevel,
+    Module,
+    Symbol,
+    Symbol,
+}
+@zero_adjoint MinimalCtx Tuple{
+    typeof(Core._call_latest),
+    typeof(CoreLogging.handle_message),
+    Any,
+    Base.CoreLogging.LogLevel,
+    String,
+    Module,
+    Symbol,
+    Symbol,
+    String,
+    Int64,
+}
+# specialized case for Builtin primitive Core._call_latest rrule for CoreLogging.handle_message kwargs call. 
+@is_primitive MinimalCtx Tuple{
+    typeof(Core._call_latest),
```

Comment:
Quick question, why do we use `Core._call_latest` here but not in the other `kwcall` [rule](https://github.com/chalk-lab/Mooncake.jl/pull/669/files#diff-41cb2b3ecbee8eea05fcec16a70a3a72284a7db1a6139741658417e1a56e7c3bR38)?